File: lib/java/opentoken/src/test/java/com/truveta/opentoken/MetadataTest.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.Map;

import org.junit.jupiter.api.Test;

class MetadataTest {

    @Test
    void testInitializeOnly() {
        Metadata metadata = new Metadata();
        Map<String, Object> result = metadata.initialize();

        assertTrue(result.containsKey(Metadata.JAVA_VERSION));
        assertTrue(result.containsKey(Metadata.PLATFORM));
        assertTrue(result.containsKey(Metadata.OPENTOKEN_VERSION));

        assertFalse(result.containsKey(Metadata.HASHING_SECRET_HASH));
        assertFalse(result.containsKey(Metadata.ENCRYPTION_SECRET_HASH));

        assertEquals(Metadata.PLATFORM_JAVA, result.get(Metadata.PLATFORM));
        assertEquals(Metadata.DEFAULT_VERSION, result.get(Metadata.OPENTOKEN_VERSION));
    }

    @Test
    void testAddHashedSecretWithHashingSecret() {
        Metadata metadata = new Metadata();
        metadata.initialize();

        String hashingSecret = "test-hashing-secret";
        Map<String, Object> result = metadata.addHashedSecret(Metadata.HASHING_SECRET_HASH, hashingSecret);

        assertTrue(result.containsKey(Metadata.HASHING_SECRET_HASH));
        assertFalse(result.containsKey(Metadata.ENCRYPTION_SECRET_HASH));
        assertNotNull(result.get(Metadata.HASHING_SECRET_HASH));
    }

    @Test
    void testAddHashedSecretWithEncryptionKey() {
        Metadata metadata = new Metadata();
        metadata.initialize();

        String encryptionKey = "test-encryption-key";
        Map<String, Object> result = metadata.addHashedSecret(Metadata.ENCRYPTION_SECRET_HASH, encryptionKey);

        assertFalse(result.containsKey(Metadata.HASHING_SECRET_HASH));
        assertTrue(result.containsKey(Metadata.ENCRYPTION_SECRET_HASH));
        assertNotNull(result.get(Metadata.ENCRYPTION_SECRET_HASH));
    }

    @Test
    void testAddHashedSecretWithBothSecrets() {
        Metadata metadata = new Metadata();
        metadata.initialize();

        String hashingSecret = "test-hashing-secret";
        String encryptionKey = "test-encryption-key";

        metadata.addHashedSecret(Metadata.HASHING_SECRET_HASH, hashingSecret);
        Map<String, Object> result = metadata.addHashedSecret(Metadata.ENCRYPTION_SECRET_HASH, encryptionKey);

        assertTrue(result.containsKey(Metadata.HASHING_SECRET_HASH));
        assertTrue(result.containsKey(Metadata.ENCRYPTION_SECRET_HASH));
        assertNotNull(result.get(Metadata.HASHING_SECRET_HASH));
        assertNotNull(result.get(Metadata.ENCRYPTION_SECRET_HASH));

        // Verify hashes are different for different inputs
        assertNotEquals(result.get(Metadata.HASHING_SECRET_HASH),
                result.get(Metadata.ENCRYPTION_SECRET_HASH));
    }

    @Test
    void testAddHashedSecretWithNullSecrets() {
        Metadata metadata = new Metadata();
        metadata.initialize();

        metadata.addHashedSecret(Metadata.HASHING_SECRET_HASH, null);
        Map<String, Object> result = metadata.addHashedSecret(Metadata.ENCRYPTION_SECRET_HASH, null);

        assertFalse(result.containsKey(Metadata.HASHING_SECRET_HASH));
        assertFalse(result.containsKey(Metadata.ENCRYPTION_SECRET_HASH));
    }

    @Test
    void testAddHashedSecretWithEmptySecrets() {
        Metadata metadata = new Metadata();
        metadata.initialize();

        metadata.addHashedSecret(Metadata.HASHING_SECRET_HASH, "");
        Map<String, Object> result = metadata.addHashedSecret(Metadata.ENCRYPTION_SECRET_HASH, "");

        assertFalse(result.containsKey(Metadata.HASHING_SECRET_HASH));
        assertFalse(result.containsKey(Metadata.ENCRYPTION_SECRET_HASH));
    }

    @Test
    void testAddHashedSecretWithCustomKey() {
        Metadata metadata = new Metadata();
        metadata.initialize();

        String customKey = "CustomSecretHash";
        String customSecret = "my-custom-secret";
        Map<String, Object> result = metadata.addHashedSecret(customKey, customSecret);

        assertTrue(result.containsKey(customKey));
        assertNotNull(result.get(customKey));
        assertEquals(Metadata.calculateSecureHash(customSecret), result.get(customKey));
    }

    @Test
    void testCalculateSecureHashWithValidInput() {
        String input = "test-input";
        String hash = Metadata.calculateSecureHash(input);

        assertNotNull(hash);
        assertFalse(hash.isEmpty());
        assertEquals(64, hash.length()); // SHA-256 produces 64 character hex string

        // Verify the hash is consistent
        String hash2 = Metadata.calculateSecureHash(input);
        assertEquals(hash, hash2);
    }

    @Test
    void testCalculateSecureHashWithKnownValue() {
        // Test with a known SHA-256 value to ensure compatibility
        String input = "hello";
        String expectedHash = "2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824";

        String actualHash = Metadata.calculateSecureHash(input);
        assertEquals(expectedHash, actualHash);
    }

    @Test
    void testCalculateSecureHashWithDifferentInputs() {
        String input1 = "input1";
        String input2 = "input2";

        String hash1 = Metadata.calculateSecureHash(input1);
        String hash2 = Metadata.calculateSecureHash(input2);

        assertNotEquals(hash1, hash2);
    }

    @Test
    void testCalculateSecureHashWithNullInput() {
        String hash = Metadata.calculateSecureHash(null);
        assertNull(hash);
    }

    @Test
    void testCalculateSecureHashWithEmptyInput() {
        String hash = Metadata.calculateSecureHash("");
        assertNull(hash);
    }

    @Test
    void testCalculateSecureHashWithUnicodeInput() {
        String input = "こんにちは"; // Japanese "hello"
        String hash = Metadata.calculateSecureHash(input);

        assertNotNull(hash);
        assertEquals(64, hash.length());

        // Verify UTF-8 encoding produces consistent results
        String hash2 = Metadata.calculateSecureHash(input);
        assertEquals(hash, hash2);
    }

    @Test
    void testMetadataConstants() {
        // Verify that the new constants are properly defined
        assertNotNull(Metadata.ENCRYPTION_SECRET_HASH);
        assertNotNull(Metadata.HASHING_SECRET_HASH);

        assertEquals("EncryptionSecretHash", Metadata.ENCRYPTION_SECRET_HASH);
        assertEquals("HashingSecretHash", Metadata.HASHING_SECRET_HASH);
    }

    @Test
    void testHashCalculationExceptionCreation() {
        String message = "Test message";
        Exception cause = new RuntimeException("Test cause");

        Metadata.HashCalculationException exception = new Metadata.HashCalculationException(message, cause);

        assertEquals(message, exception.getMessage());
        assertEquals(cause, exception.getCause());
    }
}

File: lib/python/opentoken/src/test/opentoken/metadata_test.py
"""
Copyright (c) Truveta. All rights reserved.
"""

from opentoken.metadata import Metadata, HashCalculationException


class TestMetadata:

    def test_initialize_only(self):
        metadata = Metadata()
        result = metadata.initialize()

        assert Metadata.PYTHON_VERSION in result
        assert Metadata.PLATFORM in result
        assert Metadata.OPENTOKEN_VERSION in result

        assert Metadata.HASHING_SECRET_HASH not in result
        assert Metadata.ENCRYPTION_SECRET_HASH not in result

        assert result[Metadata.PLATFORM] == Metadata.PLATFORM_PYTHON
        assert result[Metadata.OPENTOKEN_VERSION] == Metadata.DEFAULT_VERSION

    def test_add_hashed_secret_with_hashing_secret(self):
        metadata = Metadata()
        metadata.initialize()

        hashing_secret = "test-hashing-secret"
        result = metadata.add_hashed_secret(Metadata.HASHING_SECRET_HASH, hashing_secret)

        assert Metadata.HASHING_SECRET_HASH in result
        assert Metadata.ENCRYPTION_SECRET_HASH not in result
        assert result[Metadata.HASHING_SECRET_HASH] is not None

    def test_add_hashed_secret_with_encryption_key(self):
        metadata = Metadata()
        metadata.initialize()

        encryption_key = "test-encryption-key"
        result = metadata.add_hashed_secret(Metadata.ENCRYPTION_SECRET_HASH, encryption_key)

        assert Metadata.HASHING_SECRET_HASH not in result
        assert Metadata.ENCRYPTION_SECRET_HASH in result
        assert result[Metadata.ENCRYPTION_SECRET_HASH] is not None

    def test_add_hashed_secret_with_both_secrets(self):
        metadata = Metadata()
        metadata.initialize()

        hashing_secret = "test-hashing-secret"
        encryption_key = "test-encryption-key"

        metadata.add_hashed_secret(Metadata.HASHING_SECRET_HASH, hashing_secret)
        result = metadata.add_hashed_secret(Metadata.ENCRYPTION_SECRET_HASH, encryption_key)

        assert Metadata.HASHING_SECRET_HASH in result
        assert Metadata.ENCRYPTION_SECRET_HASH in result
        assert result[Metadata.HASHING_SECRET_HASH] is not None
        assert result[Metadata.ENCRYPTION_SECRET_HASH] is not None

        # Verify hashes are different for different inputs
        assert result[Metadata.HASHING_SECRET_HASH] != result[Metadata.ENCRYPTION_SECRET_HASH]

    def test_add_hashed_secret_with_null_secrets(self):
        metadata = Metadata()
        metadata.initialize()

        metadata.add_hashed_secret(Metadata.HASHING_SECRET_HASH, None)
        result = metadata.add_hashed_secret(Metadata.ENCRYPTION_SECRET_HASH, None)

        assert Metadata.HASHING_SECRET_HASH not in result
        assert Metadata.ENCRYPTION_SECRET_HASH not in result

    def test_add_hashed_secret_with_empty_secrets(self):
        metadata = Metadata()
        metadata.initialize()

        metadata.add_hashed_secret(Metadata.HASHING_SECRET_HASH, "")
        result = metadata.add_hashed_secret(Metadata.ENCRYPTION_SECRET_HASH, "")

        assert Metadata.HASHING_SECRET_HASH not in result
        assert Metadata.ENCRYPTION_SECRET_HASH not in result

    def test_add_hashed_secret_with_custom_key(self):
        metadata = Metadata()
        metadata.initialize()

        custom_key = "CustomSecretHash"
        custom_secret = "my-custom-secret"
        result = metadata.add_hashed_secret(custom_key, custom_secret)

        assert custom_key in result
        assert result[custom_key] is not None
        assert result[custom_key] == Metadata.calculate_secure_hash(custom_secret)

    def test_calculate_secure_hash_with_valid_input(self):
        input_str = "test-input"
        hash_result = Metadata.calculate_secure_hash(input_str)

        assert hash_result is not None
        assert len(hash_result) > 0
        assert len(hash_result) == 64  # SHA-256 produces 64 character hex string

        # Verify the hash is consistent
        hash2 = Metadata.calculate_secure_hash(input_str)
        assert hash_result == hash2

    def test_calculate_secure_hash_with_known_value(self):
        # Test with a known SHA-256 value to ensure compatibility
        input_str = "hello"
        expected_hash = "2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824"

        actual_hash = Metadata.calculate_secure_hash(input_str)
        assert actual_hash == expected_hash

    def test_calculate_secure_hash_with_different_inputs(self):
        input1 = "input1"
        input2 = "input2"

        hash1 = Metadata.calculate_secure_hash(input1)
        hash2 = Metadata.calculate_secure_hash(input2)

        assert hash1 != hash2

    def test_calculate_secure_hash_with_null_input(self):
        hash_result = Metadata.calculate_secure_hash(None)
        assert hash_result is None

    def test_calculate_secure_hash_with_empty_input(self):
        hash_result = Metadata.calculate_secure_hash("")
        assert hash_result is None

    def test_calculate_secure_hash_with_unicode_input(self):
        input_str = "こんにちは"  # Japanese "hello"
        hash_result = Metadata.calculate_secure_hash(input_str)

        assert hash_result is not None
        assert len(hash_result) == 64

        # Verify UTF-8 encoding produces consistent results
        hash2 = Metadata.calculate_secure_hash(input_str)
        assert hash_result == hash2

    def test_metadata_constants(self):
        # Verify that the new constants are properly defined
        assert Metadata.ENCRYPTION_SECRET_HASH is not None
        assert Metadata.HASHING_SECRET_HASH is not None

        assert Metadata.ENCRYPTION_SECRET_HASH == "EncryptionSecretHash"
        assert Metadata.HASHING_SECRET_HASH == "HashingSecretHash"

    def test_hash_calculation_exception_creation(self):
        message = "Test message"
        cause = RuntimeError("Test cause")

        exception = HashCalculationException(message, cause)

        assert str(exception) == message
        assert exception.cause == cause
